## **1. Research Questions introduction**



	“Quais são os fatores (género, número de ratings, e popularidade) que mais influenciam a média de avaliação dos filmes no MovieLens?”

Isto permite:
	•	Explorar a estrutura e a distribuição dos dados (EDA);
	•	Criar métricas derivadas (popularidade, número de votos, média ponderada);
	•	Usar modelos simples (regressão linear / árvore de decisão) para quantificar o impacto de cada variável;
	•	E escalar a análise — localmente em DuckDB, e facilmente replicável em cloud (Athena/BigQuery).

## **2. Pipeline**

### **2.1. Import necessary libraries and packages**

In [17]:
import polars as pl
import pandas as pd
import sys
import os

In [18]:
print(sys.executable)

#"C:\Users\SaraEstevesHenriques\AppData\Local\Programs\Python\Python313\python.exe" -m pip install polars

c:\Users\SaraEstevesHenriques\AppData\Local\Programs\Python\Python313\python.exe


In [11]:
print ("This is the file directory:", os.getcwd())
parquet_path = os.path.join(os.getcwd(), "moviedetails.parquet")

print("Python executable:", sys.executable)
print("Parquet exists:", os.path.exists(parquet_path))


This is the file directory: c:\Users\SaraEstevesHenriques\Documents\GitHub\BDF25_7\.git\BDF25_7\BDF25_7_data\big_data\ml-32m
Python executable: c:\Users\SaraEstevesHenriques\AppData\Local\Programs\Python\Python313\python.exe
Parquet exists: True


### **2.1. Import dataset**

#### **2.1.1. Import parquet - small ratings version!!**

In [19]:
# Eager read (loads into memory)
try:
    df_small = pl.read_parquet(parquet_path)
    print("Eager read shape:", df.shape)
    display(df.head())
except Exception as e:
    print("Eager read failed:", e)

Eager read shape: (87585, 7)


shape: (5, 7)
┌─────────┬──────────────┬──────────────┬──────────────┬──────────────┬──────────────┬─────────────┐
│ movieId ┆ title        ┆ genres       ┆ average_rati ┆ rating_count ┆ tag_count_pe ┆ user_count_ │
│ ---     ┆ ---          ┆ ---          ┆ ng           ┆ ---          ┆ r_movie      ┆ per_movie   │
│ i64     ┆ str          ┆ str          ┆ ---          ┆ i64          ┆ ---          ┆ ---         │
│         ┆              ┆              ┆ f64          ┆              ┆ i64          ┆ i64         │
╞═════════╪══════════════╪══════════════╪══════════════╪══════════════╪══════════════╪═════════════╡
│ 1       ┆ Toy Story    ┆ Adventure|An ┆ 3.897438     ┆ 84866310     ┆ 84866310     ┆ 68997       │
│         ┆ (1995)       ┆ imation|Chil ┆              ┆              ┆              ┆             │
│         ┆              ┆ dren|C…      ┆              ┆              ┆              ┆             │
│ 186     ┆ Nine Months  ┆ Comedy|Roman ┆ 2.879732     ┆ 758406       ┆ 758406       ┆ 11491       │
│         ┆ (1995)       ┆ ce           ┆              ┆              ┆              ┆             │
│ 272     ┆ "Madness of  ┆  The (1994)" ┆ 3.745054     ┆ 761760       ┆ 761760       ┆ 7935        │
│         ┆ King George  ┆              ┆              ┆              ┆              ┆             │
│ 348     ┆ Bullets Over ┆ Comedy       ┆ 3.696894     ┆ 279140       ┆ 279140       ┆ 8210        │
│         ┆ Broadway     ┆              ┆              ┆              ┆              ┆             │
│         ┆ (1994)       ┆              ┆              ┆              ┆              ┆             │
│ 458     ┆ Geronimo: An ┆ Drama|Wester ┆ 3.155412     ┆ 12972        ┆ 12972        ┆ 1081        │
│         ┆ American     ┆ n            ┆              ┆              ┆              ┆             │
│         ┆ Legend (…    ┆              ┆              ┆              ┆              ┆             │
└─────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴─────────────┘

#### **2.1.2. Import cvs - results obtained and exported from query in aws**

In [20]:
#the cvs is on this folder and its called moviedetails csv
df = pl.read_csv("moviedetails.csv")
print("CSV read successfully. Shape:", df.shape)
display(df.head())


CSV read successfully. Shape: (87585, 7)


shape: (5, 7)
┌─────────┬──────────────┬──────────────┬──────────────┬──────────────┬──────────────┬─────────────┐
│ movieId ┆ title        ┆ genres       ┆ average_rati ┆ rating_count ┆ tag_count_pe ┆ user_count_ │
│ ---     ┆ ---          ┆ ---          ┆ ng           ┆ ---          ┆ r_movie      ┆ per_movie   │
│ i64     ┆ str          ┆ str          ┆ ---          ┆ i64          ┆ ---          ┆ ---         │
│         ┆              ┆              ┆ f64          ┆              ┆ i64          ┆ i64         │
╞═════════╪══════════════╪══════════════╪══════════════╪══════════════╪══════════════╪═════════════╡
│ 20      ┆ Money Train  ┆ Action|Comed ┆ 2.871458     ┆ 589922       ┆ 589922       ┆ 4306        │
│         ┆ (1995)       ┆ y|Crime|Dram ┆              ┆              ┆              ┆             │
│         ┆              ┆ a|Thri…      ┆              ┆              ┆              ┆             │
│ 171     ┆ Jeffrey      ┆ Comedy|Drama ┆ 3.593487     ┆ 27132        ┆ 27132        ┆ 1428        │
│         ┆ (1995)       ┆              ┆              ┆              ┆              ┆             │
│ 315     ┆ Specialist,  ┆ Action|Drama ┆ 2.893674     ┆ 102717       ┆ 102717       ┆ 11413       │
│         ┆ The (1994)   ┆ |Thriller    ┆              ┆              ┆              ┆             │
│ 331     ┆ Tom & Viv    ┆ Drama        ┆ 3.289855     ┆ 1656         ┆ 1656         ┆ 552         │
│         ┆ (1994)       ┆              ┆              ┆              ┆              ┆             │
│ 386     ┆ S.F.W.       ┆ Drama        ┆ 2.822281     ┆ 5655         ┆ 5655         ┆ 377         │
│         ┆ (1994)       ┆              ┆              ┆              ┆              ┆             │
└─────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴─────────────┘

### **2.2. Data Preparation**

2.	Calcular:

•nº total de utilizadores, filmes e avaliações - não sei se o número de users é preciso
•distribuição de rating (média, mediana, desvio padrão)
•nº de filmes por género
•nº médio de ratings por filme e por utilizador

3.	Visualizar:
	
•histograma das classificações
•top 10 géneros mais avaliados
•relação entre nº de ratings e média por filme


#### 2.1.1. Titles analysis

The titles in most cases have the actual title and the movie year. The first step is to split both of this informations. 

In [21]:
df.head(20)

shape: (20, 7)
┌─────────┬──────────────┬──────────────┬──────────────┬──────────────┬──────────────┬─────────────┐
│ movieId ┆ title        ┆ genres       ┆ average_rati ┆ rating_count ┆ tag_count_pe ┆ user_count_ │
│ ---     ┆ ---          ┆ ---          ┆ ng           ┆ ---          ┆ r_movie      ┆ per_movie   │
│ i64     ┆ str          ┆ str          ┆ ---          ┆ i64          ┆ ---          ┆ ---         │
│         ┆              ┆              ┆ f64          ┆              ┆ i64          ┆ i64         │
╞═════════╪══════════════╪══════════════╪══════════════╪══════════════╪══════════════╪═════════════╡
│ 20      ┆ Money Train  ┆ Action|Comed ┆ 2.871458     ┆ 589922       ┆ 589922       ┆ 4306        │
│         ┆ (1995)       ┆ y|Crime|Dram ┆              ┆              ┆              ┆             │
│         ┆              ┆ a|Thri…      ┆              ┆              ┆              ┆             │
│ 171     ┆ Jeffrey      ┆ Comedy|Drama ┆ 3.593487     ┆ 27132        ┆ 27132        ┆ 1428        │
│         ┆ (1995)       ┆              ┆              ┆              ┆              ┆             │
│ 315     ┆ Specialist,  ┆ Action|Drama ┆ 2.893674     ┆ 102717       ┆ 102717       ┆ 11413       │
│         ┆ The (1994)   ┆ |Thriller    ┆              ┆              ┆              ┆             │
│ 331     ┆ Tom & Viv    ┆ Drama        ┆ 3.289855     ┆ 1656         ┆ 1656         ┆ 552         │
│         ┆ (1994)       ┆              ┆              ┆              ┆              ┆             │
│ 386     ┆ S.F.W.       ┆ Drama        ┆ 2.822281     ┆ 5655         ┆ 5655         ┆ 377         │
│         ┆ (1994)       ┆              ┆              ┆              ┆              ┆             │
│ …       ┆ …            ┆ …            ┆ …            ┆ …            ┆ …            ┆ …           │
│ 1233    ┆ Boot, Das    ┆ Action|Drama ┆ 4.129334     ┆ 14334228     ┆ 14334228     ┆ 15804       │
│         ┆ (Boat, The)  ┆ |War         ┆              ┆              ┆              ┆             │
│         ┆ (1981)       ┆              ┆              ┆              ┆              ┆             │
│ 1545    ┆ Ponette      ┆ Drama        ┆ 3.849922     ┆ 7716         ┆ 7716         ┆ 643         │
│         ┆ (1996)       ┆              ┆              ┆              ┆              ┆             │
│ 1772    ┆ Blues        ┆ Action|Comed ┆ 2.544862     ┆ 135149       ┆ 135149       ┆ 3143        │
│         ┆ Brothers     ┆ y|Musical    ┆              ┆              ┆              ┆             │
│         ┆ 2000 (1998)  ┆              ┆              ┆              ┆              ┆             │
│ 1783    ┆ Palmetto     ┆ Crime|Drama| ┆ 2.87548      ┆ 28622        ┆ 28622        ┆ 1301        │
│         ┆ (1998)       ┆ Mystery|Roma ┆              ┆              ┆              ┆             │
│         ┆              ┆ nce|Th…      ┆              ┆              ┆              ┆             │
│ 1801    ┆ Man in the   ┆ Action|Adven ┆ 3.106722     ┆ 497835       ┆ 497835       ┆ 7215        │
│         ┆ Iron Mask,   ┆ ture|Drama   ┆              ┆              ┆              ┆             │
│         ┆ The (199…    ┆              ┆              ┆              ┆              ┆             │
└─────────┴──────────────┴──────────────┴──────────────┴──────────────┴──────────────┴─────────────┘

In [23]:
# Extract year and remove it from title
df = df.with_columns([
    # Extract year, the 4 numbers inside the "". If no year is found it will be null
    pl.col("title").str.extract(r'\((\d{4})\)', 1).cast(pl.Int64).alias("year"),
    # Remove year from title (removes the pattern " (YYYY)" or "(YYYY)")
    pl.col("title").str.replace(r'\s*\(\d{4}\)', '').str.strip_chars('"').alias("title")
])

In [24]:
df.head()

shape: (5, 8)
┌─────────┬─────────────┬─────────────┬─────────────┬─────────────┬────────────┬────────────┬──────┐
│ movieId ┆ title       ┆ genres      ┆ average_rat ┆ rating_coun ┆ tag_count_ ┆ user_count ┆ year │
│ ---     ┆ ---         ┆ ---         ┆ ing         ┆ t           ┆ per_movie  ┆ _per_movie ┆ ---  │
│ i64     ┆ str         ┆ str         ┆ ---         ┆ ---         ┆ ---        ┆ ---        ┆ i64  │
│         ┆             ┆             ┆ f64         ┆ i64         ┆ i64        ┆ i64        ┆      │
╞═════════╪═════════════╪═════════════╪═════════════╪═════════════╪════════════╪════════════╪══════╡
│ 20      ┆ Money Train ┆ Action|Come ┆ 2.871458    ┆ 589922      ┆ 589922     ┆ 4306       ┆ 1995 │
│         ┆             ┆ dy|Crime|Dr ┆             ┆             ┆            ┆            ┆      │
│         ┆             ┆ ama|Thri…   ┆             ┆             ┆            ┆            ┆      │
│ 171     ┆ Jeffrey     ┆ Comedy|Dram ┆ 3.593487    ┆ 27132       ┆ 27132      ┆ 1428       ┆ 1995 │
│         ┆             ┆ a           ┆             ┆             ┆            ┆            ┆      │
│ 315     ┆ Specialist, ┆ Action|Dram ┆ 2.893674    ┆ 102717      ┆ 102717     ┆ 11413      ┆ 1994 │
│         ┆ The         ┆ a|Thriller  ┆             ┆             ┆            ┆            ┆      │
│ 331     ┆ Tom & Viv   ┆ Drama       ┆ 3.289855    ┆ 1656        ┆ 1656       ┆ 552        ┆ 1994 │
│ 386     ┆ S.F.W.      ┆ Drama       ┆ 2.822281    ┆ 5655        ┆ 5655       ┆ 377        ┆ 1994 │
└─────────┴─────────────┴─────────────┴─────────────┴─────────────┴────────────┴────────────┴──────┘

The following line allows to conclude that not every movie has the correspondent year set.

In [25]:
df.null_count()
#12 missing values on the year

shape: (1, 8)
┌─────────┬───────┬────────┬────────────────┬──────────────┬────────────────┬───────────────┬──────┐
│ movieId ┆ title ┆ genres ┆ average_rating ┆ rating_count ┆ tag_count_per_ ┆ user_count_pe ┆ year │
│ ---     ┆ ---   ┆ ---    ┆ ---            ┆ ---          ┆ movie          ┆ r_movie       ┆ ---  │
│ u32     ┆ u32   ┆ u32    ┆ u32            ┆ u32          ┆ ---            ┆ ---           ┆ u32  │
│         ┆       ┆        ┆                ┆              ┆ u32            ┆ u32           ┆      │
╞═════════╪═══════╪════════╪════════════════╪══════════════╪════════════════╪═══════════════╪══════╡
│ 0       ┆ 0     ┆ 0      ┆ 3153           ┆ 0            ┆ 0              ┆ 0             ┆ 615  │
└─────────┴───────┴────────┴────────────────┴──────────────┴────────────────┴───────────────┴──────┘

In [30]:
df_noyear = df.filter(pl.col("year").is_null())
df_noyear.head ()

shape: (5, 8)
┌─────────┬─────────────┬─────────────┬─────────────┬─────────────┬────────────┬────────────┬──────┐
│ movieId ┆ title       ┆ genres      ┆ average_rat ┆ rating_coun ┆ tag_count_ ┆ user_count ┆ year │
│ ---     ┆ ---         ┆ ---         ┆ ing         ┆ t           ┆ per_movie  ┆ _per_movie ┆ ---  │
│ i64     ┆ str         ┆ str         ┆ ---         ┆ ---         ┆ ---        ┆ ---        ┆ i64  │
│         ┆             ┆             ┆ f64         ┆ i64         ┆ i64        ┆ i64        ┆      │
╞═════════╪═════════════╪═════════════╪═════════════╪═════════════╪════════════╪════════════╪══════╡
│ 146796  ┆ Le strelle  ┆ (no genres  ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ null │
│         ┆ nel fosso   ┆ listed)     ┆             ┆             ┆            ┆            ┆      │
│ 156605  ┆ Paterson    ┆ (no genres  ┆ 3.712108    ┆ 177940      ┆ 177940     ┆ 1148       ┆ null │
│         ┆             ┆ listed)     ┆             ┆             ┆            ┆            ┆      │
│ 191403  ┆ New         ┆ (no genres  ┆ 3.5         ┆ 1           ┆ 0          ┆ 1          ┆ null │
│         ┆ Neighbors,  ┆ listed)     ┆             ┆             ┆            ┆            ┆      │
│         ┆ Old Fights  ┆             ┆             ┆             ┆            ┆            ┆      │
│ 195217  ┆ Star Wars:  ┆ Sci-Fi      ┆ 2.590909    ┆ 88          ┆ 88         ┆ 11         ┆ null │
│         ┆ Dresca      ┆             ┆             ┆             ┆            ┆            ┆      │
│ 205074  ┆ Falling Inn ┆ Comedy|Roma ┆ 2.7         ┆ 735         ┆ 735        ┆ 105        ┆ null │
│         ┆ Love        ┆ nce         ┆             ┆             ┆            ┆            ┆      │
└─────────┴─────────────┴─────────────┴─────────────┴─────────────┴────────────┴────────────┴──────┘

For the purpose of our analysis i think we'll need to remove the 3153 movies without average rating, these are movies without any rating

In [29]:
df_norating = df.filter(pl.col("average_rating").is_null())
df_norating.head (10)

shape: (10, 8)
┌─────────┬─────────────┬─────────────┬─────────────┬─────────────┬────────────┬────────────┬──────┐
│ movieId ┆ title       ┆ genres      ┆ average_rat ┆ rating_coun ┆ tag_count_ ┆ user_count ┆ year │
│ ---     ┆ ---         ┆ ---         ┆ ing         ┆ t           ┆ per_movie  ┆ _per_movie ┆ ---  │
│ i64     ┆ str         ┆ str         ┆ ---         ┆ ---         ┆ ---        ┆ ---        ┆ i64  │
│         ┆             ┆             ┆ f64         ┆ i64         ┆ i64        ┆ i64        ┆      │
╞═════════╪═════════════╪═════════════╪═════════════╪═════════════╪════════════╪════════════╪══════╡
│ 101237  ┆ 9500        ┆ Documentary ┆ null        ┆ 0           ┆ 4          ┆ 0          ┆ 2009 │
│         ┆ Liberty     ┆             ┆             ┆             ┆            ┆            ┆      │
│ 109647  ┆ Bengazi     ┆ Adventure|C ┆ null        ┆ 0           ┆ 4          ┆ 0          ┆ 1955 │
│         ┆             ┆ rime|Drama  ┆             ┆             ┆            ┆            ┆      │
│ 110822  ┆ Super Fly   ┆ Action|Crim ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 1973 │
│         ┆ T.N.T.      ┆ e|Drama     ┆             ┆             ┆            ┆            ┆      │
│ 116594  ┆ Driftin'    ┆ Western     ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 1946 │
│         ┆ River       ┆             ┆             ┆             ┆            ┆            ┆      │
│ 121395  ┆ Mother is a ┆ Comedy      ┆ null        ┆ 0           ┆ 3          ┆ 0          ┆ 1949 │
│         ┆ Freshman    ┆             ┆             ┆             ┆            ┆            ┆      │
│ 123794  ┆ Killer by   ┆ Crime|Thril ┆ null        ┆ 0           ┆ 3          ┆ 0          ┆ 1972 │
│         ┆ Night       ┆ ler         ┆             ┆             ┆            ┆            ┆      │
│ 129988  ┆ Tiffany     ┆ Action|Roma ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 1967 │
│         ┆ Memorandum  ┆ nce         ┆             ┆             ┆            ┆            ┆      │
│ 133497  ┆ Goldface il ┆ Adventure|C ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 1967 │
│         ┆ fantastico  ┆ rime        ┆             ┆             ┆            ┆            ┆      │
│         ┆ Superma…    ┆             ┆             ┆             ┆            ┆            ┆      │
│ 138658  ┆ I Put a Hit ┆ Comedy|Roma ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 2014 │
│         ┆ On You      ┆ nce|Thrille ┆             ┆             ┆            ┆            ┆      │
│         ┆             ┆ r           ┆             ┆             ┆            ┆            ┆      │
│ 139375  ┆ Après la    ┆ (no genres  ┆ null        ┆ 0           ┆ 1          ┆ 0          ┆ 2014 │
│         ┆ bataille    ┆ listed)     ┆             ┆             ┆            ┆            ┆      │
└─────────┴─────────────┴─────────────┴─────────────┴─────────────┴────────────┴────────────┴──────┘

#### 2.1.2. Genres

There are movies with multiple genres and others with no gender indicated "(no genres listed)". The first action here is to treat the no genres listed as nulls. One possible approach to split the genders is use explode funcion, creating a line per movie per gender.

In [31]:
#Tranform no (no genres list) to null values
df = df.with_columns(
    pl.when(pl.col("genres").str.to_lowercase().str.strip_chars() == "(no genres listed)")
    .then(None)
    .otherwise(pl.col("genres"))
    .alias("genres")
    )

In [34]:
df_nogenre = df.filter(pl.col("genres").is_null())
df_nogenre.head (10)

shape: (10, 8)
┌─────────┬──────────────┬────────┬──────────────┬──────────────┬─────────────┬─────────────┬──────┐
│ movieId ┆ title        ┆ genres ┆ average_rati ┆ rating_count ┆ tag_count_p ┆ user_count_ ┆ year │
│ ---     ┆ ---          ┆ ---    ┆ ng           ┆ ---          ┆ er_movie    ┆ per_movie   ┆ ---  │
│ i64     ┆ str          ┆ str    ┆ ---          ┆ i64          ┆ ---         ┆ ---         ┆ i64  │
│         ┆              ┆        ┆ f64          ┆              ┆ i64         ┆ i64         ┆      │
╞═════════╪══════════════╪════════╪══════════════╪══════════════╪═════════════╪═════════════╪══════╡
│ 123989  ┆ Spike of     ┆ null   ┆ 4.5          ┆ 26           ┆ 26          ┆ 2           ┆ 1988 │
│         ┆ Bensonhurst  ┆        ┆              ┆              ┆             ┆             ┆      │
│ 128393  ┆ Highway      ┆ null   ┆ 3.5          ┆ 21           ┆ 21          ┆ 3           ┆ 1977 │
│         ┆ Racer        ┆        ┆              ┆              ┆             ┆             ┆      │
│ 131320  ┆ Fort Yuma    ┆ null   ┆ 3.5          ┆ 1            ┆ 1           ┆ 1           ┆ 1955 │
│ 132896  ┆ Pelota       ┆ null   ┆ 3.5          ┆ 2            ┆ 2           ┆ 2           ┆ 1983 │
│ 137140  ┆ Blue Ridge   ┆ null   ┆ 5.0          ┆ 1            ┆ 0           ┆ 1           ┆ 2012 │
│ 139375  ┆ Après la     ┆ null   ┆ null         ┆ 0            ┆ 1           ┆ 0           ┆ 2014 │
│         ┆ bataille     ┆        ┆              ┆              ┆             ┆             ┆      │
│ 142128  ┆ The Missing  ┆ null   ┆ 1.5          ┆ 1            ┆ 0           ┆ 1           ┆ 1945 │
│         ┆ Corpse       ┆        ┆              ┆              ┆             ┆             ┆      │
│ 143217  ┆ Middle Age   ┆ null   ┆ 3.5          ┆ 2            ┆ 2           ┆ 2           ┆ 1980 │
│         ┆ Crazy        ┆        ┆              ┆              ┆             ┆             ┆      │
│ 144998  ┆ Rebels in    ┆ null   ┆ null         ┆ 0            ┆ 1           ┆ 0           ┆ 1965 │
│         ┆ Canada       ┆        ┆              ┆              ┆             ┆             ┆      │
│ 146275  ┆ 11 Minutes   ┆ null   ┆ 2.852941     ┆ 17           ┆ 17          ┆ 17          ┆ 2015 │
└─────────┴──────────────┴────────┴──────────────┴──────────────┴─────────────┴─────────────┴──────┘

In [33]:

df.null_count()


shape: (1, 8)
┌─────────┬───────┬────────┬────────────────┬──────────────┬────────────────┬───────────────┬──────┐
│ movieId ┆ title ┆ genres ┆ average_rating ┆ rating_count ┆ tag_count_per_ ┆ user_count_pe ┆ year │
│ ---     ┆ ---   ┆ ---    ┆ ---            ┆ ---          ┆ movie          ┆ r_movie       ┆ ---  │
│ u32     ┆ u32   ┆ u32    ┆ u32            ┆ u32          ┆ ---            ┆ ---           ┆ u32  │
│         ┆       ┆        ┆                ┆              ┆ u32            ┆ u32           ┆      │
╞═════════╪═══════╪════════╪════════════════╪══════════════╪════════════════╪═══════════════╪══════╡
│ 0       ┆ 0     ┆ 7080   ┆ 3153           ┆ 0            ┆ 0              ┆ 0             ┆ 615  │
└─────────┴───────┴────────┴────────────────┴──────────────┴────────────────┴───────────────┴──────┘

In [38]:
df_exploded = df.with_columns(
    pl.col("genres").str.split("|")
).explode("genres")

df_exploded.head()

shape: (5, 7)
┌─────────┬───────────────────┬─────────┬────────────────┬──────────────┬───────────────────┬──────┐
│ movieId ┆ title             ┆ genres  ┆ average_rating ┆ rating_count ┆ tag_count_per_mov ┆ year │
│ ---     ┆ ---               ┆ ---     ┆ ---            ┆ ---          ┆ ie                ┆ ---  │
│ i64     ┆ str               ┆ str     ┆ f64            ┆ i64          ┆ ---               ┆ i64  │
│         ┆                   ┆         ┆                ┆              ┆ i64               ┆      │
╞═════════╪═══════════════════╪═════════╪════════════════╪══════════════╪═══════════════════╪══════╡
│ 161582  ┆ Hell or High      ┆ Crime   ┆ 3.5625         ┆ 3176         ┆ 3176              ┆ 2016 │
│         ┆ Water             ┆         ┆                ┆              ┆                   ┆      │
│ 161582  ┆ Hell or High      ┆ Drama   ┆ 3.5625         ┆ 3176         ┆ 3176              ┆ 2016 │
│         ┆ Water             ┆         ┆                ┆              ┆                   ┆      │
│ 1982    ┆ Halloween         ┆ Horror  ┆ 3.722222       ┆ 8586         ┆ 8586              ┆ 1978 │
│ 52245   ┆ Blades of Glory   ┆ Comedy  ┆ 3.088235       ┆ 2482         ┆ 2482              ┆ 2007 │
│ 52245   ┆ Blades of Glory   ┆ Romance ┆ 3.088235       ┆ 2482         ┆ 2482              ┆ 2007 │
└─────────┴───────────────────┴─────────┴────────────────┴──────────────┴───────────────────┴──────┘

Considering the purpose of the project... decide if we need to delete the movies without values

In [ ]:
#drop a linhas com null values
#df_clean = df_exploded.drop_nulls()

### **2.2. Exploratory data analysis**

### **2.3. Feature Engineering**

Criar novas variáveis:
-	n_ratings → nº de avaliações por filme
-	avg_rating → média de rating por filme
-	genre_count → nº de géneros atribuídos
-	popularity_score = log(1 + nº de ratings) × média (para normalizar popularidade)

## **3. Models**

Objetivo: medir o peso de cada fator sobre a média de avaliação.
Modelos possíveis:
	•	Regressão Linear (OLS) → prever avg_rating com base em n_ratings, genre_count e dummies dos géneros principais.
	•	Árvore de decisão / Random Forest → avaliar a importância relativa das variáveis.
Métricas: R², RMSE, importância de features.


## **4. Visualization and Anaysis**

Gráficos:
	•	Scatter n_ratings vs avg_rating (com tendência)
	•	Boxplots de avg_rating por género
	•	Importância das features no modelo
	•	Insights:
	•	Géneros com maiores médias
	•	Géneros mais populares
	•	Relação entre popularidade e qualidade percebida


## **5. Escalability and big_data**

Executar as queries principais em DuckDB e exportar para Parquet.
	•	Demonstrar (mesmo que parcialmente) a execução de consultas analíticas escaláveis:

SELECT genre, COUNT(*) AS n_filmes, AVG(rating) AS avg_rating
FROM ratings
JOIN movies USING(movieId)
GROUP BY genre;

	•	Mostrar que a mesma lógica pode ser usada em Athena / BigQuery com datasets maiores (ex: MovieLens 25M).

## **5. Conclusions**

	•	Identificar que fatores explicam melhor as boas avaliações.
	•	Propor extensão:
	•	Modelo preditivo por utilizador (colaborativo)
	•	Integração com tags para enriquecer o conteúdo
